Attributes on functions supplied by decorators trigger `no-member`

### Bug description

```python
from functools import cache


class SomeClass:
    # Yes, the order of decorators is correct.
    @classmethod
    @cache
    def __cached_fun(cls, arg: int) -> str:  # no-member
        return str(arg)

    @classmethod
    def cache_clear(cls):
        cls.__cached_fun.cache_clear()
```


### Configuration

_No response_

### Command used

```shell
pylint a.py
```


### Pylint output

```python
a.py:13:8: E1101: Method '__cached_fun' has no 'cache_clear' member (no-member)
```


### Expected behavior

No message above. Actually everything works as expected at runtime.

### Pylint version

```shell
pylint 2.17.4
astroid 2.15.4
Python 3.11.6 (main, Oct  3 2023, 00:00:00) [GCC 13.2.1 20230728 (Red Hat 13.2.1-1)]
```


### OS / Environment

Fedora 38

### Additional dependencies

_No response_

--- a/doc/whatsnew/fragments/9246.false_positive
+++ b/doc/whatsnew/fragments/9246.false_positive
@@ -1,0 +1,4 @@
+Avoid false positives for ``no-member`` involving function
+attributes supplied by decorators.
+
+Closes #9246

--- a/pylint/checkers/typecheck.py
+++ b/pylint/checkers/typecheck.py
@@ -1060,7 +1060,7 @@ accessed. Python regular expressions are accepted.",
     def visit_delattr(self, node: nodes.DelAttr) -> None:
         self.visit_attribute(node)
 
-    # pylint: disable = too-many-branches
+    # pylint: disable = too-many-branches, too-many-statements
     @only_required_for_messages("no-member", "c-extension-no-member")
     def visit_attribute(
         self, node: nodes.Attribute | nodes.AssignAttr | nodes.DelAttr
@@ -1128,6 +1128,12 @@ accessed. Python regular expressions are accepted.",
             except astroid.DuplicateBasesError:
                 continue
             except astroid.NotFoundError:
+                # Avoid false positive in case a decorator supplies member.
+                if (
+                    isinstance(owner, (astroid.FunctionDef, astroid.BoundMethod))
+                    and owner.decorators
+                ):
+                    continue
                 # This can't be moved before the actual .getattr call,
                 # because there can be more values inferred and we are
                 # stopping after the first one which has the attribute in question.

--- a/tests/functional/n/no/no_member_decorator.py
+++ b/tests/functional/n/no/no_member_decorator.py
@@ -1,0 +1,16 @@
+"""Attributes supplied by a decorator."""
+
+from functools import lru_cache
+
+
+class SomeClass:  # pylint: disable=too-few-public-methods
+    """https://github.com/pylint-dev/pylint/issues/9246"""
+    @classmethod
+    @lru_cache
+    def __cached_fun(cls, arg: int) -> str:
+        return str(arg)
+
+    @classmethod
+    def cache_clear(cls):
+        """__cached_fun()'s @cache decorator supplies cache_clear()."""
+        cls.__cached_fun.cache_clear()

